File: 김영한 Spring/hello-spring-java/src/test/java/hello/hellospring/repository/MemoryMemberRepositoryTest.java
package hello.hellospring.repository;

import hello.hellospring.domain.Member;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import javax.swing.text.html.Option;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

class MemoryMemberRepositoryTest {

    /**
     * TEST는 메소드 각각이 서로에게 의존관계가 없어야 한다.
     */

    MemoryMemberRepository memoryMemberRepository = new MemoryMemberRepository();

    @AfterEach
    public void afterEach() {
        memoryMemberRepository.clearStore();
    }

    @Test
    public void save() {
        Member member = new Member();
        member.setName("SpringBoot!!");
        memoryMemberRepository.save(member);

        Member result = memoryMemberRepository.findById(member.getId()).get();
        // expected, actual 순으로
//        Assertions.assertEquals(member, result); 이것보다 떠 빠르게 검증하고 싶으면 아래 assertThat을 하면 된다.
        // 아래는 테스트 통과 못하면 다음 단계 못가게 막아버림
        assertThat(member).isEqualTo(result);

    }

    @Test
    public void findByName() {
        /**
         * 다른 이름을 갖는 ID 2개를 만들어보자
         */

        Member member1 = new Member();
        member1.setName("spring1");
        memoryMemberRepository.save(member1);

        // shift f6 누르면 한번에 다 바꿈
        Member member2 = new Member();
        member2.setName("spring2");
        memoryMemberRepository.save(member2);

        Member result1 = memoryMemberRepository.findByName("spring1").get();
        Member result2 = memoryMemberRepository.findByName("spring2").get();

        assertThat(result1).isEqualTo(member1);
        assertThat(result2).isEqualTo(member2);
    }

    @Test
    public void findAll() {
        Member member1 = new Member();
        member1.setName("spring1");
        memoryMemberRepository.save(member1);

        Member member2 = new Member();
        member2.setName("spring1");
        memoryMemberRepository.save(member2);

        List<Member> result = memoryMemberRepository.findAll();

        assertThat(result.size()).isEqualTo(2);

    }

}

File: 김영한 Spring/hello-spring-kotlin/src/test/kotlin/kotlinhellospring/demo/repository/MemoryMemberRepositoryTest.kt
package kotlinhellospring.demo.repository

import kotlinhellospring.demo.domain.Member
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class MemoryMemberRepositoryTest @Autowired constructor(
    private val jdbcMemberRepository: JdbcMemberRepository
) {

//    @Test
//    fun save() {
//        val member = Member()
//        memoryMemberRepository.save(member)
//        val result =  memoryMemberRepository.findById(member.id!!)
//        Assertions.assertEquals(member, result)
//    }

//    @AfterEach
//    fun afterEach() = jdbcMemberRepository.clear()

    @Test
    fun save() {
        val member = Member(name = "SpringBoot!!")
        jdbcMemberRepository.save(member)
        val result =  jdbcMemberRepository.findByName(member.name!!)
        assertThat(member).isEqualTo(result)
    }

    @Test
    fun findByName() {

        val member1 = Member(name = "spring1")
        jdbcMemberRepository.save(member1)

        val member2 = Member(name = "spring2")
        jdbcMemberRepository.save(member2)

        assertThat(member1).isEqualTo(jdbcMemberRepository.findByName("spring1"))
        assertThat(member2).isEqualTo(jdbcMemberRepository.findByName("spring2"))
    }

    @Test
    fun findAll() {
        val member1 = Member(name = "spring3")
        val member2 = Member(name = "spring4")

        jdbcMemberRepository.save(member1)
        jdbcMemberRepository.save(member2)

        val result = jdbcMemberRepository.findAll()

        assertThat(result.size).isEqualTo(2)


    }

}